### PR TITLE
[BEAM-1988] Migrate from utils.path to BFS

### PR DESCRIPTION
**Pull request by sb2nov**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

R: @chamikaramj @aaltay PTAL

### CODE REVIEW

**Comment by sb2nov** [comment]
Retest this please

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py [hunk 1]
```diff
@@ -336,10 +336,12 @@ def __init__(self, options):
     # for GCS staging locations where the potential for such clashes is high.
     if self.google_cloud_options.staging_location.startswith('gs://'):
       path_suffix = '%s.%f' % (self.google_cloud_options.job_name, time.time())
-      self.google_cloud_options.staging_location = utils.path.join(
+      filesystem = get_filesystem(self.google_cloud_options.staging_location)
+      self.google_cloud_options.staging_location = filesystem.join(
           self.google_cloud_options.staging_location, path_suffix)
-      self.google_cloud_options.temp_location = utils.path.join(
+      self.google_cloud_options.temp_location = filesystem.join(
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: What happens if `temp_location` is not a gcs location? It is an error by itself but it was handled correctly in the previous code.

I wonder if there is any value in something like `bfs.join` where `bfs` would call the right `join` method for the given type, avoiding calls to `get_file_system`.

File: sdks/python/apache_beam/runners/dataflow/internal/dependency.py [hunk 2]
```diff
@@ -284,13 +285,15 @@ def stage_job_resources(
     raise RuntimeError(
         'The --temp_location option must be specified.')
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/internal/dependency.py]: Why do we check for `temp_location` here?

Comment (sb2nov) [inline sdks/python/apache_beam/runners/dataflow/internal/dependency.py]: I'll investigate.

File: sdks/python/apache_beam/runners/dataflow/internal/apiclient.py (hunk 1 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: Yes, I think me and Cham discussed it at some point. I can work on that in another PR and clean the get_filesystem infiltration everywhere.

Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/internal/apiclient.py]: Thank you.

File: sdks/python/apache_beam/runners/dataflow/internal/dependency.py (hunk 2 above)
Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/internal/dependency.py]: Thank you does not need to happen in this PR if it is involved.

**Comment by aaltay** [comment]
This PR LGTM as is. Are you planning to make any more changes in this PR?

**Comment by sb2nov** [comment]
No the two changes we discussed seem unrelated so I can do them in separate PRs.

File: sdks/python/apache_beam/runners/dataflow/internal/dependency.py (hunk 2 above)
Comment (sb2nov) [inline sdks/python/apache_beam/runners/dataflow/internal/dependency.py]: This is just a sanity check as https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/dataflow/internal/apiclient.py#L325 would have already assigned temp_location to staging_location had it not been specified. I think we can leave it as is right now.

**Pull request closed without merging**